server/lib/bdbag/bdbag_provider.py

```diff
@@ -0,0 +1,174 @@
+import zipfile
+from typing import cast, Dict, Optional, Generator
+
+from ..import_item import ImportItem
+from ..data_map import DataMap
+from ..file_map import FileMap
+from ..entity import Entity
+from ..import_providers import ImportProvider
+
+
+_COPY_BUFSZ = 64 * 1024
+
+
+class _BagTree:
+    def __init__(self, name: str, is_dir: bool = False, url: Optional[str] = None,
+                 size: Optional[int] = None):
+        self.name = name
+        self._is_dir = is_dir
+        self.size = size
+        if is_dir:
+            self.list = {}  # type: Optional[Dict[str, _BagTree]]
+            self.url = None
+        else:
+            self.list = None
+            self.url = url
+
+    def add(self, path: pathlib.Path, url: Optional[str] = None,
+            size: Optional[int] = None) -> None:
+        self._add(path, url, path, size)
+
+    def is_dir(self) -> bool:
+        # the reason for this method is that zipfile.is_dir is a method, not a property, so
+        # it can get really confusing to mix x.is_dir with y.is_dir()
+        return self._is_dir
+
+    def _add(self, path: pathlib.Path, url: Optional[str], orig_path: pathlib.Path,
+             size: Optional[int] = None) -> None:
+        assert self.list is not None
+        name = path.parts[0]
+        if len(path.parts) == 1:
+            # leaf
+            if name in self.list:
+                raise ValueError('Duplicate entry: {}'.format(orig_path))
+            self.list[name] = _BagTree(name, url=url, size=size)
+        else:
+            if name not in self.list:
+                self.list[name] = _BagTree(name, is_dir=True)
+            dir = self.list[name]
+            if not dir.is_dir():
+                raise ValueError('Attempted to add a file where a directory '
+                                 'exists: {}'.format(orig_path))
+            dir._add(path.relative_to(name), url, orig_path, size=size)
+
+    def __str__(self) -> str:
+        s = ''
+        self._pp(s, '')
+        return s
+
+    def _pp(self, buf: str, indent: str) -> None:
+        if self.is_dir():
+            buf += indent[:-1] + '+' + self.name
+            for k, v in self.list.items():
+                v._pp(buf, indent + ' |')
+        else:
+            buf += indent + self.name
+
+
+class BDBagProvider(ImportProvider):
+    def __init__(self) -> None:
+        super().__init__('BDBag')
+
+    def lookup(self, entity: Entity) -> DataMap:
+        raise NotImplementedError('Unsupported operation')
+
+    def listFiles(self, entity: Entity) -> FileMap:
+        raise NotImplementedError('Unsupported operation')
+
+    def getDatasetUID(self, doc: object, user: object) -> str:
+        """Given a registered object, return dataset DOI"""
+        raise NotImplementedError('Unsupported operation')
+
+    def getURI(self, doc: object, user: object) -> str:
+        """Given a registered object, return a URI for it"""
+        raise NotImplementedError('Unsupported operation')
```

Comment:
All those methods should already by raising `NotImplementedError` why duplicate code?

Reply:
Probably because I forgot they do. Will fix.